xdsl/dialects/builtin.py

```diff
@@ -353,8 +353,39 @@ def print_parameter(self, printer: Printer) -> None:
                 raise ValueError(f"Invalid signedness {data}")
 
 
+class BitWidthType(TypeAttribute, ABC):
+    """
+    A type attribute with a defined bitwidth
+    """
+
+    name = "abstract.bitwidth_type"
+
+    @property
+    @abstractmethod
+    def get_bitwidth(self) -> int:
+        """
+        Returns the width of an element type in bits.
+        """
+        raise NotImplementedError()
+
+    @property
+    def get_element_width(self) -> int:
```

Comment:
```suggestion
    def get_exact_size_in_bytes(self) -> int | None:
```
Some ugly but explicit naming would be nice if we're moving this to shared infrastructure. Also, the client should decide how to handle bitwidths that are not evenly divisible by 8.

The cleanest option for me would be something like `MemoryLayout` in Swift: https://developer.apple.com/documentation/swift/memorylayout, which describes the alignment, size, and stride of a structure in memory, precisely for these kinds of calculations.

Reply:
Another option would be to go directly for this:
```
    @property
    def size(self) -> int:
        """Size of the type in bytes, defined as `ceil(bitwidth / 8)`"""
        return  bitwidth // 8 + bool(bitwidth % 8)
```

Reply:
Oh but we actually want `stride` for the calculation below!

Reply:
My understanding of `stride` is that it's the smallest multiple of `alignment` that is greater than or equal to `size`, and the alignment is target-dependent, so the type itself can only provide the size, and the rest of the logic should be in the lowering.

Reply:
I thought `alignment` would be the alignment when allocating, as in `llvm.alloca`, typically much larger than the `stride`. Or how do you see this?

Reply:
I do like the `stride` attribute, as this is essentially what everybody is using `get_exact_size_in_bytes` for.

Reply:
How about:
```
bitwidth: int
    Contiguous memory footprint in bits.
size: int | None
    Contiguous memory footprint in bytes. None if not applicable
stride: int | None
    Number of bytes from one element to the next when stored contiguously in memory. None if not applicable
```
If my understanding is correct, `alignment` is indeed target-dependent and not really a property of the type. Bitwidth must be defined, but size and stride can return None in the case of an `i4` type stored contiguously for example.

Reply:
I would say we almost always want the size to be the number of bites that fits the number of bits, so it might as well not be optional. If the client cares about the divisibility they can actually do that check themselves.

And I would drop the stride, I'm not sure how correct it is without the alignment.

Reply:
Okay, I kept it as mandatory, changed the name to `size` as this seems a clear enough destinction from bitwidth, and used in other places as well. I included a default implementation to the `ceil(bitwidth / 8)` function